src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php

```diff
@@ -146,7 +155,7 @@ public function loadChoicesForValues(array $values, $value = null)
 
         // Optimize performance in case we have an object loader and
         // a single-field identifier
-        $optimize = null === $value || is_array($value) && $value[0] === $this->idReader;
+        $optimize = is_array($value) && $this->idReader === $value[0];
```

Comment:
Even if I was the one "fixing" it in the first place and introducing a regression at the same type...

However it's clear to me now that `choice_value` in entity type should better not be used on big collections to keep this optimisation efficient.

What I think of now is that actually `$value` is always resolved by factories when stored in the cached `LazyChoiceList` which passes it to the loader each time.

What about creating another `ChoiceListInterface` depending on this logic while moving it in a `Symfony\Bridge\Doctrine\Form\ChoiceList\Factory\SingleIdFactoryDecorator` ?

Do you think it's doable to cache a map `id => value` when `$value` is not null somewhere ?

What about possibilities added by the new Cache component ?
ids should be immutable, right ? But their corresponding values may need to be updated but must not change during the lifecycle of a form since we rely on it to select choices.

How to generate a good hash based on the class and `$value` in that case ? that should not be a problem if `$value` is a string property path but if it is a closure ?

One way I see this could be implemented is by adding a `Symfony\Bridge\Doctrine\Form\ChoiceList\LazyEntityList` which would be basically a `LazyChoiceList` holding the `idReader` for mapping and using `$value` as well.

1. Current WorkFlow
-------------
1. Decorated factory creates and stores the `LazyChoiceList` storing the `DoctrineChoiceLoader` and  `$value = null`
2. `LazyChoiceList` calls the loader and passes `null` as value.
3. `DoctrineChoiceLoader` resolves `$value` with its `idReader` and load the choices
4. `DoctrineChoiceLoader` calls the decorated factory with resolved value and loaded choices
5. Decorated factory creates and stores the `ArrayChoiceList` will never call it back.
6. `DoctrineChoiceLoader` stores the `ArrayChoiceList` and will never call it back.
7. `LazyChoiceList` stores the `ArrayChoiceList` while storing the loader.

Also, if another `EntityType` is used in the same form, another `DoctrineChoiceLoader` is created by the `DoctrineType` and the same process happens again.

When the same `em`, `class` and 'query_builder` options are used for the `EntityType` the same loader is return, so the factory returns the same `LazyChoiceList` storing the loaded `ArrayChoiceList`.

2. Workflow a bit improved by this PR

Removes line 5 from above.

3. Ideal Workflow
---------
1. IdReader decorated factory resolves `$value`, `$idReader` and `$entityLoader` (instead of DoctrineType) then creates a `LazyEntityList` storing the `DoctrineChoiceLoader` and the resolved arguments, 
2. `LazyEntityList` calls the loader and passes it the resolved arguments.
3. `DoctrineChoiceLoader` load the choices with optimisation
4. `DoctrineChoiceLoader` creates an `EntityList` with an extra map ids by values
5. `DoctrineChoiceLoader` stores the `EntityList`.

What do you think ? Thanks.